src/frequenz/sdk/timeseries/_periodic_feature_extractor.py

```diff
@@ -0,0 +1,414 @@
+    buffer of historical data.
+
+    The profile is created out of all windows that are fully contained in the
+    underlying buffer with the same start and end time modulo a fixed period.
+
+    Note:
+        The oldest window or the window that is currently overwritten in the
+        `MovingWindow` is not considered in the profile.
+
+    Note:
+        When constructing a `PeriodicFeatureExtractor` object the
+        `MovingWindow` size has to be a integer multiple of the period.
+
+    Example:
+        ```python
+        from datetime import datetime, timedelta, timezone
+
+        moving_window = MovingWindow(
+            size=timedelta(days=35),
+            resampled_data_recv=microgrid.logical_meter().grid_power.new_receiver(),
+            input_sampling_period=timedelta(seconds=1),
+        )
+
+        feature_extractor = PeriodicFeatureExtractor(
+            moving_window = moving_window,
+            period=timedelta(days=7),
+        )
+
+        now = datetime.now(timezone.utc)
+
+        # create a daily weighted average for the next 24h
+        avg_24h = feature_extractor.avg(
+            now,
+            now + timedelta(hours=24),
+            weights=[0.1, 0.2, 0.3, 0.4]
+        )
+
+        # create a daily average for Thursday March 23 2023
+        th_avg_24h = feature_extractor.avg(datetime(2023, 3, 23), datetime(2023, 3, 24))
+        ```
+    """
+
+    def __init__(
+        self,
+        moving_window: MovingWindow,
+        period: timedelta,
+    ) -> None:
+        """
+        Initialize a PeriodicFeatureExtractor object.
+
+        Args:
+            moving_window: The MovingWindow that is used for the average calculation.
+            period: The distance between two succeeding intervals.
+
+        Raises:
+            ValueError: If the MovingWindow size is not a integer multiple of the period.
+        """
+        self._moving_window = moving_window
+
+        self._sampling_period = self._moving_window.sampling_period
+        """The sampling_period as float to use it for indexing of samples."""
+
+        self._period = int(period / self._sampling_period)
+        """Distance between two succeeding intervals in samples."""
+
+        _logger.debug("Initializing PeriodicFeatureExtractor!")
+        _logger.debug("MovingWindow size: %i", len(self._moving_window))
+        _logger.debug(
+            "Period between two succeeding intervals (in samples): %i",
+            self._period,
+        )
+
+        if not len(self._moving_window) % self._period == 0:
+            raise ValueError(
+                "The MovingWindow size is not a integer multiple of the period."
+            )
+
+        if not is_close_to_zero(self._period - period / self._sampling_period):
+            _logger.warning(
+                "The period is not a multiple of the sampling period. "
+                "This might result in unexpected behaviour."
+            )
```

Comment:
To be honest I haven't fully thought these cases through but I think it'll work. Since I didn't want to add any unnecessary restrictions I left the warning and decided to add follow up unit tests to get some more insights.

Reply:
I think if we don't know how it behaves we should raise, we can still change it when we have more insights.

Reply:
Fair enough! Will change to raise :).